docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+BP002
+#####
+
+:Number: BP002
+:Title: Requirements resolver
+:Author: Willian Rampazzo <willianr@redhat.com>
+:Discussions-To: https://github.com/avocado-framework/avocado/issues/3455
+:Reviewers:
+:Created: 27-Jan-2020
+:Type: Architecture Blueprint
+:Status: Draft
+
+.. contents:: Table of Contents
+
+TL;DR
+*****
+
+The current management of test assets is handled manually by the test
+developer. It is usual to have a set of repetitive code blocks to define the
+name, location, and other attributes of an asset, download it or raise an
+exception if a problem occurred and the download failed.
+
+Based on use cases compiled from the discussion on qemu-devel mailing-list [1]
+and discussions during avocado meetings, this blueprint describes the
+architecture of a requirements resolver aiming the extensibility and
+flexibility when handling different types of assets, like a file, a cloud
+image, a package, a Git repository, source codes or Operating System parameters.
+
+Motivation
+**********
+
+Implement a test that gathers the requirements during its execution may lead to
+a wrong interpretation of the test results if a requirement is not satisfied.
+The failure of a test because of a missing requirement does not mean the test
+itself failed. During its execution, the test has never reached the core test
+code; still, it may be considered a failing test.
+
+Fulfill all the test requirements beforehand can be an efficient way to handle
+requirements problems and can improve the trustworthiness of the test result.
+It means that if a test ran and failed, the code responsible for the failure is
+related to the core test and not with one of its requirements.
+
+Regardless of how the test defines a requirement, an architecture capable of
+identifying them is beneficial. Storing its references and delegating to the
+code responsible for handling each different type of requirement makes the
+overall architecture of Avocado and the requirement definition of a test more
+flexible.
+
+This blueprint discusses the architecture of a requirements resolver and
+defines the APIs for the modules responsible for handling the different
+requirements types.
+
+Specification
+*************
+
+Basics
+======
+
+A resolver is responsible for creating resolutions from a given representation.
```

Comment:
I agree that the requirements representation should not differ at all in the implementation of this blueprint. Maybe, in the future, we find a use case that needs a different format. But, basically, in the implementation of this blueprint, a JSON requirements file or a docstring, in the end, will have the same form of a dictionary representing requirements, given the specificity of each format as you pointed below, in another comment.

The idea I wanted to express in the text about a resolver is of a module that, as you said, gathers well defined standard formats as inputs and converts them to parameters for each module responsible for the requirements type.

The word **_converts_** here means receive a dictionary like:

`{"type": "file", "uri": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}`

And call a class File(**kwargs) without the `type` key and value.

Maybe the text needs a more precise explanation.

Now, not related to your comment, the conversion and class instantiation described here brought to my attention the fact that when the resolver instantiates the corresponding type class, it should also call the method that will _execute_ the main operation of that class. Having an interface that is inherited by the type modules will help with the resolver flexibility and ensure the modules supported by the resolver are compatible with it.

Let me know if what I mean here is clear or if I misunderstood what you have pointed in your comment.

Reply:
One more thing I was thinking, maybe what we are defining in this blueprint will be an Avocado plugin that runs during the job pre-test phase, and we simply call it _requirements_.

Reply:
IMO the fact that, based on dictionary, we'll end up calling a class is an implementation detail that doesn't need to propagate all the way up into the design level (which should attempt to set the concepts and its names).

Having said that, choosing the right names and wording is *very* hard, so I'm not saying that there's no conversion or resolution going on, but I do have the impression that the most important operation happening there, which IMO should be signaled to users, is that this component will take care of finding/gathering the defined requirements.  This also reinforces the idea that calling a class afterwards, is actually more closely related to the fulfillment of the requirements, rather than this resolving/gathering.

Think of something like `--prepare-and-save-job-but-dont-run-it` kind of option.  It should make you realize that during such execution, the requirements will be resolved/gathered.  Then, with an option such as `--run-job-previously-saved` (which is like Today's `replay`), it would be necessary to *fulfill* those requirements, but not to resolve/gather them again.

This is similar in concept to:
 * `avocado nlist --writes-recipes-to-dir=$PATH -- $REFERENCE` *resolving* everything about a test
 * `avocado runnable-run-recipe $PATH/01.json` *fulfilling* the execution of the "test"

So maybe, what you're calling "resolution" is a combination of what I'm calling "gathering" + "fulfillment".  I think having the diagrams would help to clear this up.

Reply:
Okay, I got your point, but I could not conclude if I should explain in detail the gathering + fulfillment behavior of the module in the current text or if I should not use the word "resolver" in this text at all and change it to "gathering" and "fulfillment" actions.

Personally, I would stick with the first option, thinking about extending the module to a full resolver in a far future and avoiding a name change in the structure.

Just one comment about the objective of a blueprint, I think it is essential to define some implementation details, like API definitions, call definitions, to make sure what was approved in the blueprint do not get misrepresented. It does not mean we should define all APIs or behavior, but having some defined helps avoid misunderstanding during the implementation.

Reply:
> Okay, I got your point, but I could not conclude if I should explain in detail the gathering + fulfillment behavior of the module in the current text or if I should not use the word "resolver" in this text at all and change it to "gathering" and "fulfillment" actions.
> 

I think it's up to you... It's just natural that we see things slightly different and this leads to thinking of different names.  The important thing is, despite the names we give to components, that they are put in the right places, and perform the right actions.  So, I'd be more concerned with, say, the diagram showing how each component will play its part, then with the exact name we give them.

> Personally, I would stick with the first option, thinking about extending the module to a full resolver in a far future and avoiding a name change in the structure.
> 

Fair enough.

> Just one comment about the objective of a blueprint, I think it is essential to define some implementation details, like API definitions, call definitions, to make sure what was approved in the blueprint do not get misrepresented. It does not mean we should define all APIs or behavior, but having some defined helps avoid misunderstanding during the implementation.

I agree only to the point that, if we're also planning the delivery of an API, then of course a high level view of the API could/should be discussed.  My point was that for the `requirements.txt` and `requirements.py` and docstring examples, the user has no API to interact with, so the translation from a dictionary to a class/function call is not relevant at all.

Reply:
Okay, got it, thanks for clarifying it!